lib/streamlit/runtime/caching/cached_message_replay.py

```diff
@@ -301,19 +285,15 @@ def save_element_message(
             return
         if len(self._cached_message_stack) >= 1:
             id_to_save = self.select_dg_to_save(invoked_dg_id, used_dg_id)
-            # Arrow dataframes have an ID but only set it when used as data editor
-            # widgets, so we have to check that the ID has been actually set to
-            # know if an element is a widget.
-            if isinstance(element_proto, Widget) and element_proto.id:
-                wid = element_proto.id
-                # TODO replace `Message` with a more precise type
-                if not self._registered_metadata:
-                    _LOGGER.error(
-                        "Trying to save widget message that wasn't registered. This should not be possible."
-                    )
-                    raise AttributeError
+            if (
+                hasattr(element_proto, "id")
+                and element_proto.id
+                and self._registered_metadata
```

Comment:
should we perhaps move this into a util function and define this as the official way to "detect" a widget? In the "Don't let fragments write outside its container path" [PR here](https://github.com/streamlit/streamlit/pull/8756/files#diff-1732b92fbcf34c3fde20854b7edc951b4f8850df372c6378829fea27462f2a32R299-R323) I have created a test with a hand-crafted list of widgets; maybe there is a good way to get all elements of Streamlit and whether or not they are widgets in the future

Reply:
Since we will likely remove the widget replay functionality soon, I would probably just leave this as.

Reply:
Alright, but please add a comment and link to the removal conversation / issue so that its not lost

Reply:
Added a comment that widget replay is deprecated (which is prepared in [this PR](https://github.com/streamlit/streamlit/pull/8817)).

Reply:
Great, thank you! Could you please link the PR in the comment as well? This kind of linking helps to connect the dots across systems; if the PR is not merged for any reason or the place in the code is forgotten to be removed, we have a dangling comment and someone else looking at this would need to start an investigation of what "will be removed soon" means. If the PR is linked, its much more clear 🙂

Reply:
Or even better, we can already merge  https://github.com/streamlit/streamlit/pull/8817 and have it deprecated :) But I will also add it to the comment.